src/lib.rs

```diff
@@ -179,94 +177,52 @@ impl<T: AsRef<str>> Grid<T> {
+        // If it exceeds terminal's width, return, since it is impossible to fit.
+        if widest_column > self.options.width {
+            return Dimensions {
+                num_lines: self.cells.len(),
+                widths: vec![self.widest_cell_width],
+            };
         }
-        // Instead of numbers of columns, try to find the fewest number of *lines*
-        // that the output will fit in.
-        let mut smallest_dimensions_yet = None;
-        for num_lines in (1..=theoretical_max_num_lines).rev() {
-            // The number of columns is the number of cells divided by the number
-            // of lines, *rounded up*.
-            let num_columns = div_ceil(self.cells.len(), num_lines);
-
-            // Early abort: if there are so many columns that the width of the
-            // *column separators* is bigger than the width of the screen, then
-            // don’t even try to tabulate it.
-            // This is actually a necessary check, because the width is stored as
-            // a usize, and making it go negative makes it huge instead, but it
-            // also serves as a speed-up.
-            let total_separator_width = (num_columns - 1) * self.options.filling.width();
-            if maximum_width < total_separator_width {
-                continue;
-            }
 
-            // Remove the separator width from the available space.
-            let adjusted_width = maximum_width - total_separator_width;
+        // Calculate number of columns with widest column size.
+        let max_num_columns = self.options.width / widest_column;
+
+        // Caculate approximate number of lines and columns.
+        let appr_num_lines = div_ceil(self.cells.len(), max_num_columns);
+        let appr_num_columns = div_ceil(self.cells.len(), appr_num_lines);
+
+        // This is a potential dimension, which can definitely fit all of the cells.
+        let mut potential_dimension = self.compute_dimensions(appr_num_lines, appr_num_columns);
+        // If all of the cells can be fit on one line, return.
+        if appr_num_lines == 1 {
+            return potential_dimension;
+        }
 
-            let potential_dimensions = self.compute_dimensions(num_lines, num_columns);
-            if potential_dimensions.widths.iter().sum::<usize>() <= adjusted_width {
-                smallest_dimensions_yet = Some(potential_dimensions);
+        // Try to increase number of columns, to see if new dimension can still fit.
+        for num_columns in appr_num_columns + 1.. {
+            let new_width = self.options.width - (num_columns - 1) * self.options.filling.width();
+            let num_lines = div_ceil(self.cells.len(), num_columns);
+            let new_dimension = self.compute_dimensions(num_lines, num_columns);
+            if new_dimension.widths.iter().sum::<usize>() <= new_width {
+                potential_dimension = new_dimension;
             } else {
                 break;
             }
```

Comment:
This break is not 100% correct. This test will fail with it. I've got a solution to that on my commit, but it will increase the computation time a bit again.
```rust
#[test]
fn weird_column_edge_case() {
    // Here, 5 columns fit while fewer columns don't. So if we exit too early
    // while increasing columns, we don't find the optimal solution.
    let grid = Grid::new(
        vec!["0", "1", "222222222", "333333333", "4", "5", "6", "7", "8"],
        GridOptions {
            direction: Direction::TopToBottom,
            filling: Filling::Spaces(2),
            width: 21,
        },
    );

    let expected = "\
        0  222222222  4  6  8\n\
        1  333333333  5  7\n\
    ";

    println!("{grid}");
    assert_eq!(grid.to_string(), expected);
}
```

Reply:
Yeah, I see the point. That's exactly the edge case I didn't think about. But I think I come up with the solution to this problem. Can I push this test with the fix commit, so you can check it out?

Reply:
Sure! I'll rebase my changes on that then

Reply:
Even tho it fixed the issue I just realized that I drastically increased the number of computations. I'll have a look into this part more later. But if you have any suggestions, just push those